idb/migration/migration.go

```diff
@@ -0,0 +1,114 @@
+package migration
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+)
+
+var DuplicateIDErr = errors.New("Duplicate ID detected.")
+
+const StatusPending = "Migration pending"
+const StatusComplete = "Migrations Complete"
+const StatusActivePrefix = "Active migration: "
+const StatusErrorPrefix = "error during migration "
+
+// A migration function should take care of writing back to metastate migration row
+type Handler func() error
+
+type Task struct {
+	MigrationId int
+
+	Handler Handler
+
+	// Description of the migration
+	Description string
+}
+
+type State struct {
+	Err     error
+	Status  string
+	Running bool
+}
+
+type Migration struct {
+	mutex sync.Mutex
+
+	state State
```

Comment:
I'll embed the State object, it's also used as a return type so I need to keep those values grouped.

Reply:
Actually, embedding wont work. It would cause the state values to be accessed without being locked.

I think that this is going to be the cleanest implementation, the alternative would be for "GetStatus" to return 4 values but I prefer to have them wrapped up in an object in case we update the state (i.e. in the future "progress" would be nice)